File: docs/algorithms/code/algorithms/findMode.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */


function findMode(root: TreeNode | null): number[] {
  const arr: number[] = [];
  let base = 0, count = 0, maxCount = 0;

  const update = (num: number) => {
    if (base === num) {
      count++;
      maxCount = Math.max(maxCount, count);
    } else {
      let prevCount = count;
      count = 1;
      base = num;
      maxCount = Math.max(maxCount, count, prevCount);
    }
  }

  const stack: TreeNode[] = [];
  while (stack.length || root) {
    while (root) {
      stack.push(root);
      root = root.left;
    }
    root = stack.pop();
    update(root.val);
    arr.push(root.val);
    root = root.right;
  }

  const res: number[] = [];

  base = arr[0];
  count = 1;

  console.log('maxCount', maxCount);
  if (maxCount === 1) {
    res.push(base);
  }

  for (let i = 1; i < arr.length; i++) {
    const val = arr[i];
    if (val === base) {
      count++;
      if (count === maxCount) {
        res.push(val);
      }
    } else {
      count = 1;
      base = val
      if (maxCount === 1) {
        res.push(val);
      }
    }
  }

  return res;
};

File: docs/algorithms/code/algorithms/findMode.go
// 思路：遍历二叉树，用哈希表记录每个节点的值出现的次数，然后遍历哈希表，找到出现次数最多的节点的值，返回
// 复杂度分析：时间复杂度 O(n)，空间复杂度 O(n)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func findMode(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	max := 0
	hash := map[int]int{}
	stack := []*TreeNode{}
	for len(stack) > 0 || root != nil {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		hash[root.Val]++
		if hash[root.Val] > max {
			max = hash[root.Val]
		}
    root = root.Right
	}
	res := []int{}
	for k, v := range hash {
		if v == max {
			res = append(res, k)
		}
	}
	return res
}